### ISSUE TITLE
Intellisense suggestions is very slow with scikit learn package

### ISSUE THREAD

**Issue by YachaywasiAI** [microsoft/pylance-release#3908]
- VSCode Version: 1.75.0
- OS Version: PopOs 22.04 LTS
- VSCode Python extension v2023.2.0
- VSCode Pylance extension v2023.2.10  

Steps to Reproduce:
1. Clean installation of VSCode, with only Material Icon and the last Python extensions.
2. Miniconda environment with packages Scikit-learn 1.12, Pytorch 1.13.1
3. The following code causes slow suggestions to intellisense (specific when I import scikit-learn package)
4. https://gist.github.com/YachaywasiAI/1ac6b0511f828b487de4ef25f1eab0ea
5. When I type last line, the parameters inside of function roc_auc_score. The time to suggest word "labels" or "predictions" is high.

**Comment by debonte** [comment]
Appears to have regressed in 2022.8.41.

Repro steps:
1. pip install scikit-learn==1.2.1, torch==1.13.1, segmentation-models-pytorch==0.3.2
2. Start with code shown below.
3. Wait for Pylance to initialize and indexing to finish.
4. After `"auc_score":`, type `roc_auc_score(` -- feel free to use the completion
5. Type `pre` and wait for completion list. It takes about 2 seconds on my machine. Sometimes it's fast, but then if you delete `pre` and try again it will be slow.

2022.8.41:
![scikit-learn](https://user-images.githubusercontent.com/12820898/216740428-808ed6ae-5bd7-4164-93b2-8a7c4a3bc5a9.gif)

2022.8.40:
![scikit-learn-working](https://user-images.githubusercontent.com/12820898/216740433-6526275e-1acc-4cc3-a993-6dcd2dcc145a.gif)

```python
from segmentation_models_pytorch.metrics import (
    get_stats,
    f1_score,
    iou_score,
)

from sklearn.metrics import roc_auc_score

import torch


def compute_metrics(
    labels,
    raw_predictions,
    criterion,
    threshold=0.5,
    reduction='macro'
):
    predictions = torch.sigmoid(raw_predictions)
    stats = get_stats(
        predictions,
        labels.type(torch.uint8),
        threshold=threshold,
        mode='binary'
    )

    return {
        'loss': criterion(raw_predictions, labels).item(),
        'f1_score': f1_score(*stats, reduction=reduction).item(),
        'iou_score': iou_score(*stats, reduction=reduction).item(),
        "auc_score": 
        #'auc_score': roc_auc_score(labels.view(-1), predictions.view(-1)) # THIS LINE
    }
```

**Comment by erictraut** [comment]
Strange. I'm not able to repro the problem on my machine. The code snippet (with the line uncommented) completes full analysis in 95-125ms. Completions for "pre" also appear almost instantly. I've tried with both Pyright and Pylance, and I see the same behavior in terms of performance.

@debonte, are you able to repro the slow behavior with the Pyright extension?

**Comment by debonte** [comment]
> @debonte, are you able to repro the slow behavior with the Pyright extension?

Yes, I'm able to repro this with Pyright 1.1.292.

I thought it might have something to do with the many other packages installed in my test project's venv, but it also repros for me in a new project with only the three packages above pip installed.

> The code snippet (with the line uncommented)...

Note that if you uncomment that line (containing `roc_auc_score`) and then start Pyright/Pylance, the issue doesn't repro. Follow the steps above where `roc_auc_score` is initially not used in the file (beyond the import).

**Comment by heejaechang** [comment]
is it with "python.analysis.useLibraryCodeForTypes: true"?

**Comment by debonte** [comment]
> is it with "python.analysis.useLibraryCodeForTypes: true"?

@heejaechang, thanks, that didn't occur to me. Yes, it only repros for me with `useLibraryCodeForTypes` enabled, and I had it explicitly enabled so Pyright was running that way also.

**Comment by heejaechang** [comment]
@erictraut another `sklearn` perf issue around `getFunctionInferredReturnType` and  `isAfterNodeReachable`

minimal repro step is

```python
from sklearn.metrics import roc_auc_score

roc_auc_score(None, None)  
```

running checker will take about 5 seconds. or you can type `.` after `roc_auc_score(None, None)` to bring up a completion which will take about the same time.

```
(19192) Background analysis message: analyze
(19192) [BG(1)] runWithCancellationToken ...
(19192) [BG(1)]   analyzing: c:\pyTemp\junk\test1.py ...
(19192) [BG(1)]     parsing: c:\pyTemp\junk\test1.py (0ms)
(19192) [BG(1)]     binding: c:\pyTemp\junk\test1.py (1ms)
(19192) [BG(1)]     checking: c:\pyTemp\junk\test1.py ...
(19192) [BG(1)]       getType ...
(19192) [BG(1)]         _getFunctionInferredReturnType ...
(19192) [BG(1)]           _getFunctionInferredReturnType ...
(19192) [BG(1)]             _getFunctionInferredReturnType [Function '_unique' (sklearn.utils._encode)] (60ms)
(19192) [BG(1)]             _getFunctionInferredReturnType [Function '_encode' (sklearn.utils._encode)] (545ms)
(19192) [BG(1)]             isAfterNodeReachable ...
(19192) [BG(1)]               isAfterNodeReachable ...
(19192) [BG(1)]                 getTypeOfExpression ...
(19192) [BG(1)]                   getTypeOfExpression ...
(19192) [BG(1)]                     getTypeOfExpression ...
(19192) [BG(1)]                       getTypeOfExpression ...
(19192) [BG(1)]                         isAfterNodeReachable ...
(19192) [BG(1)]                           getTypeOfExpression ...
(19192) [BG(1)]                             getTypeOfExpression ...
(19192) [BG(1)]                               evaluateTypeForSubnode ...
(19192) [BG(1)]                                 evaluateTypesForStatement ["arr" (scipy.sparse._sputils) [183:9]] (335ms)
(19192) [BG(1)]                               evaluateTypeForSubnode ["arr" (scipy.sparse._sputils) [183:9]] (335ms)
(19192) [BG(1)]                             getTypeOfExpression ["arr.max" (scipy.sparse._sputils) [190:30]] (337ms)
(19192) [BG(1)]                           getTypeOfExpression ["arr.min" (scipy.sparse._sputils) [191:30]] (489ms)
(19192) [BG(1)]                         isAfterNodeReachable [function '"get_index_dtype"  [147:5]' (scipy.sparse._sputils) [147:1]] (489ms)
(19192) [BG(1)]                         evaluateTypeForSubnode ...
(19192) [BG(1)]                           evaluateTypesForStatement ...
(19192) [BG(1)]                             getTypeOfExpression ["idx_dtype" (scipy.sparse._compressed) [1017:19]] (59ms)
(19192) [BG(1)]                             evaluateTypeForSubnode ...
(19192) [BG(1)]                               evaluateTypesForStatement ["nnzs" (scipy.sparse._compressed) [1016:9]] (57ms)
(19192) [BG(1)]                             evaluateTypeForSubnode ["nnzs" (scipy.sparse._compressed) [1016:9]] (57ms)
(19192) [BG(1)]                           evaluateTypesForStatement ["self.indices" (scipy.sparse._compressed) [1198:13]] (213ms)
(19192) [BG(1)]                         evaluateTypeForSubnode ["self.indices" (scipy.sparse._compressed) [1198:13]] (213ms)
(19192) [BG(1)]                         getTypeOfExpression ...
(19192) [BG(1)]                           isAfterNodeReachable ...
(19192) [BG(1)]                             getTypeOfExpression ...
(19192) [BG(1)]                               getTypeOfExpression ...
(19192) [BG(1)]                                 getTypeOfExpression ["M.nonzero" (scipy.sparse._coo) [190:38]] (130ms)
(19192) [BG(1)]                                 getTypeOfExpression ...
(19192) [BG(1)]                                   evaluateTypeForSubnode ...
(19192) [BG(1)]                                     evaluateTypesForStatement ["summed" (scipy.sparse._coo) [528:9]] (180ms)
(19192) [BG(1)]                                   evaluateTypeForSubnode ["summed" (scipy.sparse._coo) [528:9]] (180ms)
(19192) [BG(1)]                                   _getFunctionInferredReturnType [Function 'to_native' (scipy.sparse._sputils)] (651ms)
(19192) [FG] parsing: c:\pyTemp\junk\test1.py (0ms)
(19192) [FG] binding: c:\pyTemp\junk\test1.py (0ms)
(19192) [BG(1)]                                 getTypeOfExpression ["mask.all" (scipy.sparse._coo) [311:20]] (1963ms)
(19192) [BG(1)]                               getTypeOfExpression ["self.data.astype" (scipy.sparse._coo) [195:25]] (2178ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: getTypeOfExpression (2178ms)
(19192) [BG(1)]                             getTypeOfExpression ["self._check" (scipy.sparse._coo) [197:9]] (2210ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: getTypeOfExpression (2210ms)
(19192) [BG(1)]                           isAfterNodeReachable [function '"__init__"  [127:9]' (scipy.sparse._coo) [127:5]] (2210ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: isAfterNodeReachable (2210ms)
(19192) [BG(1)]                         getTypeOfExpression ["self._set_self" (scipy.sparse._compressed) [55:21]] (2211ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: getTypeOfExpression (2211ms)
(19192) [BG(1)]                       getTypeOfExpression ["self.indices.max" (scipy.sparse._compressed) [96:33]] (3144ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: getTypeOfExpression (3144ms)
(19192) [BG(1)]                     getTypeOfExpression ["self._swap" (scipy.sparse._compressed) [100:47]] (3144ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: getTypeOfExpression (3144ms)
(19192) [BG(1)]                     _getFunctionInferredReturnType [Function 'to_native' (scipy.sparse._sputils)] (183ms)
(19192) [BG(1)]                   getTypeOfExpression ["self.data.astype" (scipy.sparse._compressed) [104:25]] (3353ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: getTypeOfExpression (3353ms)
(19192) [BG(1)]                   _getFunctionInferredReturnType ...
(19192) [BG(1)]                     getTypeOfExpression ...
(19192) [BG(1)]                       getTypeOfExpression ...
(19192) [BG(1)]                         getTypeOfExpression ...
(19192) [BG(1)]                           evaluateTypeForSubnode ...
(19192) [BG(1)]                             evaluateTypesForStatement ["self.indices" (scipy.sparse._compressed) [159:9]] (88ms)
(19192) [BG(1)]                           evaluateTypeForSubnode ["self.indices" (scipy.sparse._compressed) [159:9]] (88ms)
(19192) [BG(1)]                         getTypeOfExpression ["self.indices.max" (scipy.sparse._compressed) [186:20]] (88ms)
(19192) [BG(1)]                       getTypeOfExpression ["self.indices.min" (scipy.sparse._compressed) [189:20]] (89ms)
(19192) [BG(1)]                     getTypeOfExpression ["np.diff(self.indptr).min" (scipy.sparse._compressed) [192:20]] (99ms)
(19192) [BG(1)]                   _getFunctionInferredReturnType [Function 'check_format' (scipy.sparse._compressed)] (99ms)
(19192) [BG(1)]                 getTypeOfExpression ["self.check_format" (scipy.sparse._compressed) [106:9]] (3453ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: getTypeOfExpression (3453ms)
(19192) [BG(1)]               isAfterNodeReachable [function '"__init__"  [26:9]' (scipy.sparse._compressed) [26:5]] (3453ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: isAfterNodeReachable (3453ms)
(19192) [BG(1)]             isAfterNodeReachable [function '"label_binarize"  [425:5]' (sklearn.preprocessing._label) [425:1]] (3453ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: isAfterNodeReachable (3453ms)
(19192) [BG(1)]             _getFunctionInferredReturnType ...
(19192) [BG(1)]               getTypeOfExpression ...
(19192) [BG(1)]                 _getFunctionInferredReturnType [Function '__getitem__' (scipy.sparse._index)] (70ms)
(19192) [BG(1)]               getTypeOfExpression ["Y.getcol" (sklearn.preprocessing._label) [589:17]] (297ms)
(19192) [BG(1)]             _getFunctionInferredReturnType [Function 'label_binarize' (sklearn.preprocessing._label)] (308ms)
(19192) [BG(1)]             _getFunctionInferredReturnType ...
(19192) [BG(1)]               getTypeOfExpression ["average_weight.sum" (sklearn.metrics._base) [98:23]] (52ms)
(19192) [BG(1)]             _getFunctionInferredReturnType [Function '_average_binary_score' (sklearn.metrics._base)] (55ms)
(19192) [BG(1)]           _getFunctionInferredReturnType [Function '_multiclass_roc_auc_score' (sklearn.metrics._ranking)] (4425ms)
[Info  - 12:28:58 AM] (19192) [BG(1)] Long operation: _getFunctionInferredReturnType (4425ms)
(19192) [BG(1)]           _getFunctionInferredReturnType [Function '_binary_roc_auc_score' (sklearn.metrics._ranking)] (117ms)
(19192) [BG(1)]         _getFunctionInferredReturnType [Function 'roc_auc_score' (sklearn.metrics._ranking)] (4601ms)
[Info  - 12:28:59 AM] (19192) [BG(1)] Long operation: _getFunctionInferredReturnType (4601ms)
(19192) [BG(1)]       getType ["roc_auc_score(None, None)" (test1) [3:1]] (4601ms)
[Info  - 12:28:59 AM] (19192) [BG(1)] Long operation: getType (4601ms)
(19192) [BG(1)]     checking: c:\pyTemp\junk\test1.py (4603ms)
[Info  - 12:28:59 AM] (19192) [BG(1)] Long operation: checking: c:\pyTemp\junk\test1.py (4603ms)
(19192) [BG(1)]   analyzing: c:\pyTemp\junk\test1.py (4604ms)
[Info  - 12:28:59 AM] (19192) [BG(1)] Long operation: analyzing: c:\pyTemp\junk\test1.py (4604ms)
(19192) [BG(1)] runWithCancellationToken (4604ms)
[Info  - 12:28:59 AM] (19192) [BG(1)] Long operation: runWithCancellationToken (4604ms)
```

**Comment by heejaechang** [comment]
related to https://github.com/microsoft/pylance-release/issues/3978 and https://github.com/microsoft/pylance-release/issues/4016

**Comment by heejaechang** [comment]
merging all scikit-learn perf related issues to https://github.com/microsoft/pylance-release/issues/3978
